### PR TITLE
ROC-5545 hearing questions save to DM

### PR DESCRIPTION
**Pull request by karthikkannankainos**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/ROC-5545

### Change description ###
Save Hearing Questions for claimant and defendant and upload to DM

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/cmc/claimstore/documents/content/directionsquestionnaire/HearingContentProvider.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+    private static final String DISABLED_ACCESS = "Disabled Access";
+    private static final String YES = "YES";
+    private static final String NO = "NO";
+
+    private Function<ExpertReport, ExpertReportContent> mapExpertReport = report ->
+        ExpertReportContent.builder().expertName(report.getExpertName())
+            .expertReportDate(Formatting.formatDate(report.getExpertReportDate()))
+            .build();
+    private Function<UnavailableDate, String> mapToISOFullStyle = unavailableDate ->
```
Comment (simad) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/documents/content/directionsquestionnaire/HearingContentProvider.java]: maybe I am missing something, but why can't these functions be simple methods? ie. `String mapToISOFullStyle(UnavailableDate unavailableDate)`? 😳 (and similarly for mapExpertReport)

File: src/main/java/uk/gov/hmcts/cmc/claimstore/documents/content/directionsquestionnaire/HearingContentProvider.java [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        Optional.ofNullable(unavailableDate)
+            .map(UnavailableDate::getUnavailableDate)
+            .map(DateUtils::toISOFullStyle).orElse("");
+
+    private BiConsumer<RequireSupport, HearingContent.HearingContentBuilder> mapSupportRequirement =
+        (support, builder) -> {
```
Comment (simad) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/documents/content/directionsquestionnaire/HearingContentProvider.java]: I don't particularly like this "building by side effect" pattern, but I understand we have it everywhere so it's ok :D

File: domain-sample-data/src/main/java/uk/gov/hmcts/cmc/domain/models/sampledata/SampleHearingLocation.java [hunk 3]
```diff
@@ -0,0 +1,20 @@
+package uk.gov.hmcts.cmc.domain.models.sampledata;
+
+import uk.gov.hmcts.cmc.domain.models.directionsquestionnaire.CourtLocationType;
+import uk.gov.hmcts.cmc.domain.models.directionsquestionnaire.HearingLocation;
+
+import java.util.function.Supplier;
+
+public class SampleHearingLocation {
+
+    public static Supplier<HearingLocation> defaultHearingLocation = () -> HearingLocation.builder()
```
[the remark this answers is not in the document]
Comment (sabahirfan) [reply domain-sample-data/src/main/java/uk/gov/hmcts/cmc/domain/models/sampledata/SampleHearingLocation.java]: Agree `Supplier` looks unnecessary here

File: src/main/resources/citizen/templates/document/defendantResponseReceipt.html [hunk 4]
```diff
@@ -1,5 +1,5 @@
 <html>
-<head>
+<head>x
```
Comment (sabahirfan) [inline src/main/resources/citizen/templates/document/defendantResponseReceipt.html]: please remove `x`

File: src/main/java/uk/gov/hmcts/cmc/claimstore/documents/content/directionsquestionnaire/ClaimantDirectionsQuestionnaireContentProvider.java [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+            .orElseThrow(IllegalArgumentException::new);
+
+        ImmutableMap.Builder<String, Object> contentBuilder = new ImmutableMap.Builder<>();
+        contentBuilder.putAll(claimContentProvider.createContent(claim));
+        contentBuilder.put("hearingContent", hearingContentProvider.mapDirectionQuestionnaire
+            .apply(claimantDirectionsQuestionnaire));
+        claim.getClaimantRespondedAt().ifPresent(respondedAt -> {
+            contentBuilder.put("claimantSubmittedOn", formatDateTime(respondedAt));
+        });
+        contentBuilder.put("formNumber", "OCON180");
```
Comment (sabahirfan) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/documents/content/directionsquestionnaire/ClaimantDirectionsQuestionnaireContentProvider.java]: As form number is static, can be part of html template itself.

File: src/main/java/uk/gov/hmcts/cmc/claimstore/documents/content/directionsquestionnaire/HearingContentProvider.java [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+
+import static java.util.stream.Collectors.toList;
+import static uk.gov.hmcts.cmc.ccd.util.StreamUtil.asStream;
+
+@Component
+public class HearingContentProvider {
+
+    private static final String DISABLED_ACCESS = "Disabled Access";
+    private static final String YES = "YES";
+    private static final String NO = "NO";
```
Comment (sabahirfan) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/documents/content/directionsquestionnaire/HearingContentProvider.java]: Are you sure we want these as capitals in PDF? We are using `Yes` / `No` in other PDFs

File: src/main/java/uk/gov/hmcts/cmc/claimstore/events/DocumentUploadHandler.java [hunk 7]
```diff
@@ -107,6 +115,24 @@ public void uploadSettlementAgreementDocument(CountersignSettlementAgreementEven
...
     }
 
+    @EventListener
+    public void uploadClaimantDirectionsQuestionnaireToDM(ClaimantResponseEvent event) {
+        Claim claim = event.getClaim();
+
+        ResponseRejection responseRejection = claim.getClaimantResponse()
+            .filter(ResponseRejection.class::isInstance)
+            .map(ResponseRejection.class::cast)
+            .orElse(null);
```
[the remark this answers is not in the document]
Comment (karthikkannankainos) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/events/DocumentUploadHandler.java]: The event is fired for all the response. But only RejectionResponse has the direction questionnaire. If the AcceptationResponse, then we can just ignore the event. Thus done this way.

File: src/main/java/uk/gov/hmcts/cmc/claimstore/documents/content/directionsquestionnaire/ClaimantDirectionsQuestionnaireContentProvider.java (hunk 5 above)
Comment (karthikkannankainos) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/documents/content/directionsquestionnaire/ClaimantDirectionsQuestionnaireContentProvider.java]: Following the standard. All the formNumbers are passed.

File: src/main/java/uk/gov/hmcts/cmc/claimstore/documents/content/directionsquestionnaire/HearingContentProvider.java (hunk 6 above)
Comment (karthikkannankainos) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/documents/content/directionsquestionnaire/HearingContentProvider.java]: Will Sort it. I think when I moved to constant, I copy pased I think. :-) My mistake.

File: src/main/java/uk/gov/hmcts/cmc/claimstore/documents/content/directionsquestionnaire/HearingContentProvider.java (hunk 1 above)
Comment (karthikkannankainos) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/documents/content/directionsquestionnaire/HearingContentProvider.java]: Function, Supplier, Consumer are predefined interface. Its strict definition enforces input and return types. So its better to use predefined FI. Everything can be done thru normal method. But having method adhering to a interface (Definition) is much more cleaner IMHO.

File: src/main/java/uk/gov/hmcts/cmc/claimstore/documents/content/directionsquestionnaire/HearingContentProvider.java (hunk 2 above)
Comment (karthikkannankainos) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/documents/content/directionsquestionnaire/HearingContentProvider.java]: Lol. You are rt. There are places where they call a method to return consumer. :-). But I didnt like the existing approach, So I used the Biconsumer.

File: src/main/resources/citizen/templates/document/defendantResponseReceipt.html (hunk 4 above)
Comment (karthikkannankainos) [reply src/main/resources/citizen/templates/document/defendantResponseReceipt.html]: Sure.

File: src/test/java/uk/gov/hmcts/cmc/claimstore/documents/content/directionsquestionnaire/HearingContentProviderTest.java [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+
+    @Test(expected = IllegalArgumentException.class)
+    public void mapDirectionsQuestionnaireThrowsException() {
+        hearingContentProvider.mapDirectionQuestionnaire.apply(null);
+    }
+
+    @Test
+    public void mapDirectionsQuestionnaireMapsHearingContent() {
+        DirectionsQuestionnaire dq = SampleDirectionsQuestionnaire.builder().build();
+        HearingContent hearingContent = hearingContentProvider.mapDirectionQuestionnaire.apply(dq);
```
[the remark this answers is not in the document]
Comment (karthikkannankainos) [reply src/test/java/uk/gov/hmcts/cmc/claimstore/documents/content/directionsquestionnaire/HearingContentProviderTest.java]: .get() gets codacy issues. Thus using ifPresent. Also the ifPresent is used on the test object and not on the return object.

File: src/main/java/uk/gov/hmcts/cmc/claimstore/events/DocumentUploadHandler.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (karthikkannankainos) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/events/DocumentUploadHandler.java]: The Event is common for ResponseRejection and ResponseAcceptation. But the questionnaire or the event listener should perform action only if rejection. Thus did it till orElse() and did a if down there.

[the remark this answers is not in the document]
Comment (simad) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/events/DocumentUploadHandler.java]: yeah, sorry, didn't explain myself properly. The logic is perfect, I just pointed out you can avoid the null check by chaining things, ie 
``` claim.getClaimantResponse()
            .filter(ResponseRejection.class::isInstance)
            .map(ResponseRejection.class::cast)
            .flatMap(ResponseRejection::getDirectionsQuestionnaire)
            .ifPresent(dq -> {
                // do stuff with dq
            });```

but it's a stylistic choice :D

File: src/test/java/uk/gov/hmcts/cmc/claimstore/documents/content/directionsquestionnaire/HearingContentProviderTest.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (simad) [reply src/test/java/uk/gov/hmcts/cmc/claimstore/documents/content/directionsquestionnaire/HearingContentProviderTest.java]: oh, I see. It's one of those examples I think the tools get in the way, to be honest

[the remark this answers is not in the document]
Comment (karthikkannankainos) [reply src/test/java/uk/gov/hmcts/cmc/claimstore/documents/content/directionsquestionnaire/HearingContentProviderTest.java]: generally .get() without condition is also a NO NO as Optional is to avoid getting nullpointer exception. But for Test I agree its too much.

File: src/main/java/uk/gov/hmcts/cmc/claimstore/documents/content/directionsquestionnaire/HearingContentProvider.java (hunk 1 above)
Comment (simad) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/documents/content/directionsquestionnaire/HearingContentProvider.java]: I hear you, but I'd say a method definition is equivalent to a `Function`, as in it still enforces a return type and the parameters types. I'd use a Function if I needed to pass that as a parameter to another function, or store it in a map, or compose it / do higher level operations on it, but if all you're doing is calling it, I feel it's just less readable.

File: src/test/java/uk/gov/hmcts/cmc/claimstore/documents/content/directionsquestionnaire/HearingContentProviderTest.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (simad) [reply src/test/java/uk/gov/hmcts/cmc/claimstore/documents/content/directionsquestionnaire/HearingContentProviderTest.java]: yeah, I only ever use .get() in tests ;)

File: src/main/resources/citizen/templates/document/claimantDirectionsQuestionnaire.html [hunk 9]
```diff
@@ -0,0 +1,388 @@
...
+      </div></td>
+      <td><div>{{ claimant.email }}</div></td>
+    </tr>
+    </tbody>
+  </table>
+
+
+
+
+
```
Comment (sabahirfan) [inline src/main/resources/citizen/templates/document/claimantDirectionsQuestionnaire.html]: too many extra lines here

Comment (karthikkannankainos) [reply src/main/resources/citizen/templates/document/claimantDirectionsQuestionnaire.html]: Done

File: domain-sample-data/src/main/java/uk/gov/hmcts/cmc/domain/models/sampledata/SampleHearingLocation.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (karthikkannankainos) [reply domain-sample-data/src/main/java/uk/gov/hmcts/cmc/domain/models/sampledata/SampleHearingLocation.java]: Removed it. But in test data setup. Supplier can provide JIT object instead of Static objects.

File: src/main/java/uk/gov/hmcts/cmc/claimstore/documents/content/directionsquestionnaire/HearingContentProvider.java [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+
+    private BiConsumer<ExpertRequest, HearingContent.HearingContentBuilder> mapExpertRequest =
+        (expertRequest, builder) -> {
+            builder.hasExpertReport(YES);
+            builder.courtPermissionForExpertReport(YES);
+            builder.reasonWhyExpertAdvice(expertRequest.getReasonForExpertAdvice());
+            builder.expertExamineNeeded(expertRequest.getExpertEvidenceToExamine());
+        };
+
+    public Function<DirectionsQuestionnaire, HearingContent> mapDirectionQuestionnaire =
```
Comment (sabahirfan) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/documents/content/directionsquestionnaire/HearingContentProvider.java]: Implementing functional interface here and not used as the assignment target for a lambda expression or method reference, appears misuse of functional interface. This can be plane vanilla java  method. i.e. 
```
HearingContent  mapDirectionQuestionnaire(DirectionsQuestionnaire questionnaire) {
...
}
```

File: src/main/java/uk/gov/hmcts/cmc/claimstore/documents/content/FullDefenceResponseContentProvider.java [hunk 11]
```diff
@@ -70,6 +78,10 @@
         content.put("evidences", evidences);
         content.put("evidenceComment", evidenceComment);
         content.put("formNumber", DEFENCE_FORM_NO);
+
+        fullDefenceResponse.getDirectionsQuestionnaire().ifPresent(dq ->
+            content.put("hearingContent", hearingContentProvider.mapDirectionQuestionnaire.apply(dq)));
```
Comment (sabahirfan) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/documents/content/FullDefenceResponseContentProvider.java]: `mapDirectionQuestionnaire.apply` in tests is fine but in code a functional method should be used as the assignment target for a lambda expression or method reference.

File: src/main/java/uk/gov/hmcts/cmc/claimstore/documents/content/PartAdmissionResponseContentProvider.java [hunk 12]
```diff
@@ -104,6 +108,8 @@ public PartAdmissionResponseContentProvider(
         );
 
         content.put("formNumber", ADMISSIONS_FORM_NO);
+        partAdmissionResponse.getDirectionsQuestionnaire().ifPresent(dq ->
+            content.put("hearingContent", hearingContentProvider.mapDirectionQuestionnaire.apply(dq)));
```
Comment (sabahirfan) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/documents/content/PartAdmissionResponseContentProvider.java]: mapDirectionQuestionnaire.apply in tests is fine but in code a functional method should be used as the assignment target for a lambda expression or method reference.

File: src/main/java/uk/gov/hmcts/cmc/claimstore/documents/content/directionsquestionnaire/ClaimantDirectionsQuestionnaireContentProvider.java [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+            .map(ResponseRejection.class::cast)
+            .map(ResponseRejection::getDirectionsQuestionnaire)
+            .filter(Optional::isPresent)
+            .map(Optional::get)
+            .orElseThrow(IllegalArgumentException::new);
+
+        ImmutableMap.Builder<String, Object> contentBuilder = new ImmutableMap.Builder<>();
+        contentBuilder.putAll(claimContentProvider.createContent(claim));
+        contentBuilder.put("hearingContent", hearingContentProvider.mapDirectionQuestionnaire
+            .apply(claimantDirectionsQuestionnaire));
```
Comment (sabahirfan) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/documents/content/directionsquestionnaire/ClaimantDirectionsQuestionnaireContentProvider.java]: mapDirectionQuestionnaire.apply in tests is fine but in code a functional method should be used as the assignment target for a lambda expression or method reference.

File: src/main/java/uk/gov/hmcts/cmc/claimstore/documents/content/directionsquestionnaire/HearingContentProvider.java [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+        questionnaire -> {
+
+            Optional.ofNullable(questionnaire).orElseThrow(IllegalArgumentException::new);
+
+            HearingContent.HearingContentBuilder contentBuilder = HearingContent.builder();
+
+            questionnaire.getRequireSupport()
+                .ifPresent(support -> mapSupportRequirement.accept(support, contentBuilder));
```
Comment (sabahirfan) [inline src/main/java/uk/gov/hmcts/cmc/claimstore/documents/content/directionsquestionnaire/HearingContentProvider.java]: `mapSupportRequirement.accept` in tests is fine but in code a functional method should be used as the assignment target for a lambda expression or method reference.

File: src/main/java/uk/gov/hmcts/cmc/claimstore/documents/content/directionsquestionnaire/HearingContentProvider.java (hunk 2 above)
Comment (sabahirfan) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/documents/content/directionsquestionnaire/HearingContentProvider.java]: This BiConsumer is not used as the assignment target for a lambda expression or method reference. It's not a valid use case for BiConsumer here.

File: ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/domain/CaseEvent.java [hunk 15]
```diff
@@ -44,7 +44,8 @@
     ACTION_REVIEW_COMMENTS("ActionReviewComments"),
     ASSIGNING_FOR_DIRECTIONS("AssigningForDirections"),
     REFERRED_TO_MEDIATION("ReferredToMediation"),
-    DRAW_ORDER("DrawOrder");
+    DRAW_ORDER("DrawOrder"),
+    CLAIMANT_DIRECTIONS_QUESTIONNAIRE_UPLOAD("ClaimantDirectionsQuestionnaireUpload");
```
Comment (sabahirfan) [inline ccd-adapter/src/main/java/uk/gov/hmcts/cmc/ccd/domain/CaseEvent.java]: Do we have a separate PR for defendant pdf?

File: src/main/resources/citizen/templates/document/claimantDirectionsQuestionnaire.html [hunk 16]
```diff
@@ -0,0 +1,450 @@
...
+    </td>
+    <td>
+      <div>
+        {{ hearingContent.hasExpertReport }}
+      </div>
+    </td>
+  </tr>
+  {% set reportNo = 0 %}
+  {% for report in hearingContent.expertReports %}
+  {% set reportNo = reportNo + 1 %}
```
Comment (sabahirfan) [inline src/main/resources/citizen/templates/document/claimantDirectionsQuestionnaire.html]: This is not required, prefer to use `loop.index` instead

File: src/main/resources/citizen/templates/document/defendantResponseReceipt.html [hunk 17]
```diff
@@ -814,5 +858,308 @@ <h3>What happens next</h3>
...
+    </td>
+    <td>
+      <div>
+        {{ hearingContent.hasExpertReport }}
+      </div>
+    </td>
+  </tr>
+  {% set reportNo = 0 %}
+  {% for report in hearingContent.expertReports %}
+  {% set reportNo = reportNo + 1 %}
```
Comment (sabahirfan) [inline src/main/resources/citizen/templates/document/defendantResponseReceipt.html]: This is not required, prefer to use `loop.index` instead

Comment (karthikkannankainos) [reply src/main/resources/citizen/templates/document/defendantResponseReceipt.html]: Done

File: src/main/java/uk/gov/hmcts/cmc/claimstore/documents/content/directionsquestionnaire/HearingContentProvider.java (hunk 14 above)
Comment (karthikkannankainos) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/documents/content/directionsquestionnaire/HearingContentProvider.java]: I dont understand what you are trying to say. The Function is declared as target for lambda. But `.accept` to be used as assignment target ?

File: src/main/java/uk/gov/hmcts/cmc/claimstore/documents/content/directionsquestionnaire/ClaimantDirectionsQuestionnaireContentProvider.java (hunk 13 above)
Comment (karthikkannankainos) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/documents/content/directionsquestionnaire/ClaimantDirectionsQuestionnaireContentProvider.java]: I dont understand what you are trying to say. The Function is declared as target for lambda. But `.apply` to be used as assignment target ?

File: src/main/java/uk/gov/hmcts/cmc/claimstore/documents/content/directionsquestionnaire/HearingContentProvider.java (hunk 10 above)
Comment (karthikkannankainos) [reply src/main/java/uk/gov/hmcts/cmc/claimstore/documents/content/directionsquestionnaire/HearingContentProvider.java]: Yeah. I am changing these to java methods.

File: src/main/resources/citizen/templates/document/defendantResponseReceipt.html [hunk 18]
```diff
@@ -814,5 +858,306 @@ <h3>What happens next</h3>
...
+      <div>
+        {{ hearingContent.hasExpertReport }}
+      </div>
+    </td>
+  </tr>
+  {% for report in hearingContent.expertReports %}
+  <tr>
+    <td>
+      <div>
+        Report {{ loop.index }}
```
Comment (sabahirfan) [inline src/main/resources/citizen/templates/document/defendantResponseReceipt.html]: Are you sure you don't need `(loop.index +1)` ? 
Also if there is only one item in collection, normally we don't display  the number. i.e. `Report` instead of `Report 1` when size is one of the expertReports collection. Please check with P.O.

Comment (karthikkannankainos) [reply src/main/resources/citizen/templates/document/defendantResponseReceipt.html]: Hi @sabahirfan I have to push the changes. Currently merging backend requirement for downloading document as well. Soon will push the change

Comment (sabahirfan) [reply src/main/resources/citizen/templates/document/defendantResponseReceipt.html]: Great thanks 👍
